### DISCUSSION TITLE
Unable to silence warnings from the framework/pico-sdk

### DISCUSSION THREAD

**Discussion by AgainPsychoX** [earlephilhower/arduino-pico#2104] [General]
While compiling my project there are a lot of warnings coming vendor code I can't directly change, so I wan to silence it to focus on my own project files. 

### My setup

I use PlatformIO include Visual Studio Code on Windows.

I use combination of `build_unflags`/`build_src_flags` to apply warnings only to my sourced compilation units, and then extra script to change paths in `CPPPATH` to use `-isystem` to ignore the warnings.

My `platformio.ini`:
```ini
[env:pico]
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = pico
framework = arduino
board_build.core = earlephilhower
; ...
build_unflags = 
	-Wall -Wextra -Wpedantic -Wvla ; hide warnings from framework & libraries
	-std=gnu++14 -std=gnu++17
build_flags =
	-std=gnu++20
; ...
	; Explicitly add toolchain headers (list from 'arm-none-eabi-g++.exe -v') to fix VS Code IntelliSense
	; List from `arm-none-eabi-g++.exe -v`, section `#include <...> search starts here`.
	-isystem $PROJECT_CORE_DIR/packages/toolchain-gccarmnoneeabi/arm-none-eabi/include/c++/9.3.1
	-isystem $PROJECT_CORE_DIR/packages/toolchain-gccarmnoneeabi/arm-none-eabi/include/c++/9.3.1/arm-none-eabi
	-isystem $PROJECT_CORE_DIR/packages/toolchain-gccarmnoneeabi/arm-none-eabi/include/c++/9.3.1/backward
	-isystem $PROJECT_CORE_DIR/packages/toolchain-gccarmnoneeabi/lib/gcc/arm-none-eabi/9.3.1/include
	-isystem $PROJECT_CORE_DIR/packages/toolchain-gccarmnoneeabi/lib/gcc/arm-none-eabi/9.3.1/include-fixed
	-isystem $PROJECT_CORE_DIR/packages/toolchain-gccarmnoneeabi/arm-none-eabi/include
build_src_flags = 
	-Wall -Wextra -Wpedantic
extra_scripts = 
	scripts/convert_sysincludes.py ; hide warnings from framework & libraries
```

I use extra script `scripts/convert_sysincludes.py` to convert some includes to be included as `-isystem`:
```py
# Script to help hide warnings from framework & libraries
# From https://community.platformio.org/t/silence-warnings-for-dependencies-external-libraries/33387/8

from pathlib import Path

# PlatformIO injected stuff (I wish it was typed... https://github.com/platformio/platformio-docs/issues/342)
Import("env" ,"projenv") # type: ignore
env, projenv, DefaultEnvironment = env, projenv, DefaultEnvironment # type: ignore

platform = env.PioPlatform()
if env.GetProjectOption("board_build.core") == 'earlephilhower':
	FRAMEWORK_DIR = Path(platform.get_package_dir("framework-arduinopico"))
else:
	FRAMEWORK_DIR = Path(platform.get_package_dir("framework-arduino-mbed"))
framework_includes = list()
filtered_cpppath = list()

# Apply these changes to current working env, the project env and the global env
for e in (env, projenv, DefaultEnvironment()):
	for p in e["CPPPATH"]:
		# Is the current include path inside the framework directory?
		if FRAMEWORK_DIR in Path(p).parents:
			framework_includes.append(p)
		else:
			filtered_cpppath.append(p)
	e.Replace(CPPPATH=filtered_cpppath)
	e.Append(CCFLAGS=[("-isystem", p) for p in framework_includes])

```

### **Sadly, it's not enough**.

Example warnings when compiling:
```log
In file included from C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/rp2_common/hardware_adc/include/hardware/adc.h:12,
                 from src\sampling.cpp:2:
C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/rp2_common/hardware_gpio/include/hardware/gpio.h:881:33: warning: ISO C++ does not permit named variadic macros [-Wvariadic-macros]
  881 | #define CU_REGISTER_DEBUG_PINS(p...) \
      |                                 ^~~
In file included from C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/common/pico_base/include/pico/types.h:12,
                 from C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/common/pico_base/include/pico.h:24,
                 from C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/rp2_common/hardware_adc/include/hardware/adc.h:10:
C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/rp2_common/hardware_base/include/hardware/address_mapped.h: In function 'uint32_t xip_alias_check_addr(const void*)':
C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/common/pico_base/include/pico/assert.h:35:34: warning: ISO C++ forbids braced-groups within expressions [-Wpedantic]
   35 | #define valid_params_if(x, test) ({if (PARAM_ASSERTIONS_ENABLED(x)) assert(test);})
      |                                  ^
C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/rp2_common/hardware_base/include/hardware/address_mapped.h:95:5: note: in expansion of macro 'valid_params_if'
   95 |     valid_params_if(ADDRESS_ALIAS, rc >= XIP_MAIN_BASE && rc < XIP_NOALLOC_BASE);
      |    
```

Using `pio run --verbose` I can inspect final compiler arguments:
```
arm-none-eabi-g++ -o .pio\build\pico\src\sampling.cpp.o -c -std=gnu++20 -fno-exceptions -fno-rtti -isystem C:\Users\PsychoX\.platformio/packages/toolchain-gccarmnoneeabi/arm-none-eabi/include/c++/9.3.1 -isystem C:\Users\PsychoX\.platformio/packages/toolchain-gccarmnoneeabi/arm-none-eabi/include/c++/9.3.1/arm-none-eabi -isystem C:\Users\PsychoX\.platformio/packages/toolchain-gccarmnoneeabi/arm-none-eabi/include/c++/9.3.1/backward -isystem C:\Users\PsychoX\.platformio/packages/toolchain-gccarmnoneeabi/lib/gcc/arm-none-eabi/9.3.1/include -isystem C:\Users\PsychoX\.platformio/packages/toolchain-gccarmnoneeabi/lib/gcc/arm-none-eabi/9.3.1/include-fixed -isystem C:\Users\PsychoX\.platformio/packages/toolchain-gccarmnoneeabi/arm-none-eabi/include -Os -Werror=return-type -Wno-psabi -march=armv6-m -mcpu=cortex-m0plus -mthumb -ffunction-sections -fdata-sections -iprefixC:\Users\PsychoX\.platformio\packages\framework-arduinopico @C:\Users\PsychoX\.platformio\packages\framework-arduinopico\lib\platform_inc.txt -Wall -Wextra -Wpedantic -isystem C:\Users\PsychoX\.platformio\packages\framework-arduinopico\cores\rp2040 -isystem C:\Users\PsychoX\.platformio\packages\framework-arduinopico\cores\rp2040\api\deprecated -isystem C:\Users\PsychoX\.platformio\packages\framework-arduinopico\cores\rp2040\api\deprecated-avr-comp -isystem C:\Users\PsychoX\.platformio\packages\framework-arduinopico\include -isystem C:\Users\PsychoX\.platformio\packages\framework-arduinopico\variants\rpipico -isystem C:\Users\PsychoX\.platformio\packages\framework-arduinopico\libraries\EEPROM\src -isystem C:\Users\PsychoX\.platformio\packages\framework-arduinopico\libraries\LittleFS\src -isystem C:\Users\PsychoX\.platformio\packages\framework-arduinopico\libraries\SPI\src -isystem C:\Users\PsychoX\.platformio\packages\framework-arduinopico\cores\rp2040 -isystem C:\Users\PsychoX\.platformio\packages\framework-arduinopico\cores\rp2040\api\deprecated -isystem C:\Users\PsychoX\.platformio\packages\framework-arduinopico\cores\rp2040\api\deprecated-avr-comp -isystem C:\Users\PsychoX\.platformio\packages\framework-arduinopico\include -isystem C:\Users\PsychoX\.platformio\packages\framework-arduinopico\variants\rpipico -DPLATFORMIO=60114 -DARDUINO_RASPBERRY_PI_PICO -DARDUINO_ARCH_RP2040 -DUSBD_MAX_POWER_MA=500 -DUSER_SETUP_LOADED=1 -DST7796_DRIVER=1 -DTFT_WIDTH=320 -DTFT_HEIGHT=480 -DTFT_MISO=4 -DTFT_MOSI=3 -DTFT_SCLK=2 -DTFT_CS=5 -DTFT_DC=7 -DTFT_RST=6 -DTOUCH_CS=8 -DLOAD_GLCD=1 -DLOAD_FONT2=1 -DLOAD_FONT4=1 -DLOAD_FONT6=1 -DLOAD_FONT7=1 -DLOAD_FONT8=1 -DLOAD_GFXFF=1 -DSMOOTH_FONT=1 -DSPI_FREQUENCY=80000000 -DARDUINO=10810 -DARDUINO_ARCH_RP2040 -DF_CPU=133000000L -DBOARD_NAME=\"pico\" -DARM_MATH_CM0_FAMILY -DARM_MATH_CM0_PLUS -DTARGET_RP2040 -DPICO_FLASH_SIZE_BYTES=2097152 -DCFG_TUSB_MCU=OPT_MCU_RP2040 -DUSB_VID=0x2e8a -DUSB_PID=0x000a "-DUSB_MANUFACTURER=\"Raspberry Pi\"" 
-DUSB_PRODUCT=\"Pico\" -DSERIALUSB_PID=0x000a -DPICO_CYW43_ARCH_THREADSAFE_BACKGROUND=1 -DCYW43_LWIP=1 -DLWIP_IPV4=1 -DLWIP_IGMP=1 -DLWIP_CHECKSUM_CTRL_PER_NETIF=1 -DLWIP_IPV6=0 -DARDUINO_VARIANT=\"rpipico\" -Iinclude -Isrc -I.pio\libdeps\pico\CRC32\src -I.pio\libdeps\pico\TFT_eSPI src\sampling.cpp
```

And I can see `-iprefixC:\Users\PsychoX\.platformio\packages\framework-arduinopico @C:\Users\PsychoX\.platformio\packages\framework-arduinopico\lib\platform_inc.txt` which the extra script I use can do nothing about (or rather would be quite hard) to make it as system (so warnings ignored) include.

While researching issue I found #615:
https://github.com/earlephilhower/arduino-pico/blob/2a256f9c25021deee92a5eece6137dbe1438e0ec/tools/platformio-build.py#L185-L192

* Shouldn't it all be `-isystem` anyways?
* Maybe I am missing something? Is there maybe easier way to I disable the warnings from framework?

**Comment by AgainPsychoX** [comment]
### Workaround using more extra scripting

I updated my extra script as workaround, adding all the file parsing, then prefixes unrolling and changing the framework includes to use `-isystem`. Gist: https://gist.github.com/AgainPsychoX/45d55d4d7004cb6f20f7920807f72981 .

I guess it's not perfect, as now the whole process has to create temporary files because of how long the shell command is with all those arguments. Here is `pio run --verbose 2>&1 > verbose.log` after modifying one of my project files: Gist: https://gist.github.com/AgainPsychoX/0c4b897d110dfdcdb7ce9c0971be4d76#file-modified_only_one_of_my_files-log

### Workaround using other include flag

Meanwhile I learnt there is `-iwithprefixbefore` (acts like `-I`) and `-iwithprefix` (acts like `-idirafter`). The priority of includes is different, but it seems working if I change them in the `platform_inc.txt` file - which makes my script modification I just wrote useless, welp. https://gcc.gnu.org/onlinedocs/gcc/Directory-Options.html

**Do you think could it be updated in the repo?** I don't want to create pull request, since I am not sure if it changes/breaks anything due to the include order changes; would require some intuition to judge it and testing to make sure, I imagine. 

### By the way

Using those workarounds, nor any of my attempts so far, I haven't managed to get rid of all warnings. While I see them no more while modifying my files, there are still some after cleaning the environment - due to recompilation of some framework provided `.cpp` files - but in this case its rare issue. Gist of example log: https://gist.github.com/AgainPsychoX/0c4b897d110dfdcdb7ce9c0971be4d76#file-modified_only_one_of_my_files-log

**Comment by AgainPsychoX** [comment]
Update: I just found out that **using second workaround doesn't work** for me: VS Code `c_cpp_properties.json` is missing stuff if `platform_inc.txt` contains just `-iwithprefix`s, but it could work if code in `platformio-build.py` would be changed too (also to `-iwithprefix`): 

https://github.com/earlephilhower/arduino-pico/blob/2a256f9c25021deee92a5eece6137dbe1438e0ec/tools/platformio-build.py#L46-L52

For now I will be using my extra script, but I hope you consider if `-iwithprefix` could be used in the official version. I could make pull request if you want, but again: it would require knowledge/testing if changed include order affects anything.

**Comment by earlephilhower** [comment]
Those headers are used in the core and built in the CI without warnings.  Are you using a non standard compiler or set up?

**Comment by AgainPsychoX** [comment]
I don't think so.

Versions are in the logs:
```
PACKAGES: 
 - framework-arduinopico @ 1.30700.0+sha.88ccf0c (git+https://github.com/earlephilhower/arduino-pico.git#88ccf0c256dc717ff932adc9e2d84485214bf2e5) 
 - tool-rp2040tools @ 1.0.2 
 - toolchain-rp2040-earlephilhower @ 5.120300.240125 (12.3.0, https://github.com/earlephilhower/pico-quick-toolchain/releases/download/2.2.0/x86_64-w64-mingw32.arm-none-eabi-d04e724.240125.zip)
```

The example warning I included in the first post seems to come from `-Wpedantic` which got applied to code in framework/pico sdk headers (i.e. `pico/assert.h`) included by my code (my `cpp` file).

_BTW, sorry for late response._

**Comment by earlephilhower** [comment]
Moved to a discussion because now that I'm back home I can see your P.IO config is neither using the default compiler we use (gcc 12.3 from the pico-quick-toolchain repo) nor the C++ standard (`-std=gnu++17`) we build against.

Since it's not really core core related, I think your best bet is to go the the P.IO forums and see if they can figure out some magic to apply different configs to different files.

**Reply by AgainPsychoX** [reply]
> your P.IO config is neither using the default compiler we use (gcc 12.3 from the pico-quick-toolchain repo

What do you mean? It is GCC 12.3.0 from `pico-quick-toolchain`. It's in my post above (as well in logs I provided in previous posts).

I can try using C++17 if you want in a second.

**Reply by AgainPsychoX** [reply]
> Since it's not really core core related, I think your best bet is to go the the P.IO forums and see if they can figure out some magic to apply different configs to different files.

The PlatformIO has `build_src_flags` to apply flags only to settings, and it's not enough here because the core builds script marks includes as regular project includes (`-I`/`-iwithprefixbefore`) instead as system/vendor includes (`-isystem`/-iwithprefixs`). 

Sorry if I provided too much text to read, I know it might be overwhelming but I wanted to provide full view of the situation.

**Reply by earlephilhower** [reply]
I'm not a PIO expert, but the [first post](https://github.com/earlephilhower/arduino-pico/discussions/2104#discussion-6510609) has this bit:

````
...
	-isystem $PROJECT_CORE_DIR/packages/toolchain-gccarmnoneeabi/arm-none-eabi/include/c++/9.3.1
	-isystem $PROJECT_CORE_DIR/packages/toolchain-gccarmnoneeabi/arm-none-eabi/include/c++/9.3.1/arm-none-eabi
	-isystem $PROJECT_CORE_DIR/packages/toolchain-gccarmnoneeabi/arm-none-eabi/include/c++/9.3.1/backward
	-isystem $PROJECT_CORE_DIR/packages/toolchain-gccarmnoneeabi/lib/gcc/arm-none-eabi/9.3.1/include
	-isystem $PROJECT_CORE_DIR/packages/toolchain-gccarmnoneeabi/lib/gcc/arm-none-eabi/9.3.1/include-fixed
...
````

which seems to indicate you're using directories and headers from some other package with GCC 9.3.1.  Might want to double-check your setup and do full verbose dumps and see what compiler is being called exactly, and update those add'l directories (because they're not going to work too well w/GCC 12.3 I can almost guarantee).

**Reply by AgainPsychoX** [reply]
Indeed those includes are wrong, my bad. 

I fixed those to look like:
```ini
	; Explicitly add internal toolchain headers to fix VS Code IntelliSense
	; List from `echo | arm-none-eabi-g++.exe -E -Wp,-v -`, section `#include <...> search starts here`.
	; framework-arduino-mbed specific:
	-isystem $PROJECT_CORE_DIR/packages/toolchain-rp2040-earlephilhower/lib/gcc/arm-none-eabi/12.3.0/include
 	-isystem $PROJECT_CORE_DIR/packages/toolchain-rp2040-earlephilhower/lib/gcc/arm-none-eabi/12.3.0/include-fixed
	-isystem $PROJECT_CORE_DIR/packages/toolchain-rp2040-earlephilhower/arm-none-eabi/include
```

Anyways, for now I will comment those out to avoid any issues coming from that.

**Reply by AgainPsychoX** [reply]
After removing those, without my updated extra script that workarounds the issue, I still get warnings from not my code:

Exactly the same as before, or in the comments before. Examples:
```
In file included from C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/rp2_common/hardware_adc/include/hardware/adc.h:12,
                 from src\sampling.cpp:3:
C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/rp2_common/hardware_gpio/include/hardware/gpio.h:881:33: warning: ISO C++ does not permit named variadic macros [-Wvariadic-macros]
  881 | #define CU_REGISTER_DEBUG_PINS(p...) \
      |                                 ^~~
In file included from C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/common/pico_base/include/pico/types.h:12,
                 from C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/common/pico_base/include/pico.h:24,
                 from C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/rp2_common/hardware_adc/include/hardware/adc.h:10:
C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/rp2_common/hardware_base/include/hardware/address_mapped.h: In function 'uint32_t xip_alias_check_addr(const void*)':
C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/common/pico_base/include/pico/assert.h:35:34: warning: ISO C++ forbids braced-groups within expressions [-Wpedantic]
   35 | #define valid_params_if(x, test) ({if (PARAM_ASSERTIONS_ENABLED(x)) assert(test);})
      |                                  ^
C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/rp2_common/hardware_base/include/hardware/address_mapped.h:95:5: note: in expansion of macro 'valid_params_if'
   95 |     valid_params_if(ADDRESS_ALIAS, rc >= XIP_MAIN_BASE && rc < XIP_NOALLOC_BASE);
      |     ^~~~~~~~~~~~~~~
```

I am running those with `pio run --verbose 2>&1 > verbose.log` after using PIO > Clean of course. Full log again here:  https://gist.github.com/AgainPsychoX/c24d113a09475382eeb0e7545da334a5

**Reply by earlephilhower** [reply]
I'm still suspect looking at those logs.  There's no full path to the compiler, only the executable name, which might take it from your `$PATH` and not the toolchain directory.  But I'm not 100% sure so YMMV.

In any case, I might suggest starting with the documented working config with no edits, as described in the [docs](https://github.com/earlephilhower/arduino-pico/blob/master/docs/platformio.rst).  That's what CI is configured with here, and it's not dropping any warnings in the logs I see.  (Actually since P.IO and the Arduino-Build tool should be using essentially identical command lines and GCC12.3 doesn't barf w/add'l error and warning settings).  You can then add things one at a time to pinpoint the problem.

Also, this shows up in the Arduino API headers, completely outside of the core or the Pico SDK as well:
````
arm-none-eabi-g++ -o .pio\build\pico\FrameworkArduino\Bootsel.cpp.o -c -fno-exceptions -fno-rtti -std=gnu++17 -Os -Werror=return-type -Wno-psabi -march=armv6-m -mcpu=cortex-m0plus -mthumb -ffunction-sections -fdata-sections -iprefixC:\Users\PsychoX\.platformio\packages\framework-arduinopico @C:\Users\PsychoX\.platformio\packages\framework-arduinopico\lib\platform_inc.txt -DPLATFORMIO=60114 -DARDUINO_RASPBERRY_PI_PICO -DARDUINO_ARCH_RP2040 -DUSBD_MAX_POWER_MA=500 -DUSER_SETUP_LOADED=1 -DST7796_DRIVER=1 -DTFT_WIDTH=320 -DTFT_HEIGHT=480 -DTFT_MISO=4 -DTFT_MOSI=3 -DTFT_SCLK=2 -DTFT_CS=5 -DTFT_DC=7 -DTFT_RST=6 -DTOUCH_CS=8 -DLOAD_GLCD=1 -DLOAD_FONT2=1 -DLOAD_FONT4=1 -DLOAD_FONT6=1 -DLOAD_FONT7=1 -DLOAD_FONT8=1 -DLOAD_GFXFF=1 -DSMOOTH_FONT=1 -DSPI_FREQUENCY=80000000 -DARDUINO=10810 -DARDUINO_ARCH_RP2040 -DF_CPU=133000000L -DBOARD_NAME=\"pico\" -DARM_MATH_CM0_FAMILY -DARM_MATH_CM0_PLUS -DTARGET_RP2040 -DPICO_FLASH_SIZE_BYTES=2097152 -DCFG_TUSB_MCU=OPT_MCU_RP2040 -DUSB_VID=0x2e8a -DUSB_PID=0x000a "-DUSB_MANUFACTURER=\"Raspberry Pi\"" -DUSB_PRODUCT=\"Pico\" -DSERIALUSB_PID=0x000a -DPICO_CYW43_ARCH_THREADSAFE_BACKGROUND=1 -DCYW43_LWIP=1 -DLWIP_IPV4=1 -DLWIP_IGMP=1 -DLWIP_CHECKSUM_CTRL_PER_NETIF=1 -DLWIP_IPV6=0 -DARDUINO_VARIANT=\"rpipico\" -IC:\Users\PsychoX\.platformio\packages\framework-arduinopico\cores\rp2040 -IC:\Users\PsychoX\.platformio\packages\framework-arduinopico\cores\rp2040\api\deprecated -IC:\Users\PsychoX\.platformio\packages\framework-arduinopico\cores\rp2040\api\deprecated-avr-comp -IC:\Users\PsychoX\.platformio\packages\framework-arduinopico\include -IC:\Users\PsychoX\.platformio\packages\framework-arduinopico\variants\rpipico C:\Users\PsychoX\.platformio\packages\framework-arduinopico\cores\rp2040\Bootsel.cpp
In file included from c:\users\psychox\.platformio\packages\framework-arduinopico\arduinocore-api\api\PluggableUSB.h:23,
                 from c:\users\psychox\.platformio\packages\framework-arduinopico\arduinocore-api\api\arduinoapi.h:33,
                 from C:\Users\PsychoX\.platformio\packages\framework-arduinopico\cores\rp2040/api/ArduinoAPI.h:2,
                 from C:\Users\PsychoX\.platformio\packages\framework-arduinopico\cores\rp2040/Arduino.h:28:
c:\users\psychox\.platformio\packages\framework-arduinopico\arduinocore-api\api\USBAPI.h:34:12: warning: ISO C++ prohibits anonymous structs [-Wpedantic]
   34 |     struct {
      |            ^
````

 So something seems dicey in the basic configuration.  Starting clean and adding in things is probably your best bet.

**Reply by earlephilhower** [reply]
I was able to get your warnings in the SDK and Arduino by adding `-Wpedantic` to the build commands (which is not used in CI).  So one mystery solved.

In a nutshell, you can't build the SDK or anything that uses Arduino includes with `-Wpedantic` and expect a clean build.

If you have the time then sending a PR to the SDK and the Arduino.cc guys changing the headers to work would be a great idea.  OTW, you might be able to hack your local copies of the Arduino.cc headers and the Pico SDK (but I wouldn't recommend it for your sanity!).  I.e. add a `#pragma` to disable the specific warnings dumped in the build in the specific headers (bracketed with the appropriate `push` and `pop` so as not to pollute further compilation units).

**Reply by AgainPsychoX** [reply]
> I'm still suspect looking at those logs. There's no full path to the compiler, only the executable name, which might take it from your $PATH and not the toolchain directory. But I'm not 100% sure so YMMV. 

I think PlatformIO build process adds relevant compiler path to PATH variable indeed, I can verify that using extra scripts using `print(env.Dump())` which shows me a lot, including:

![image](https://github.com/earlephilhower/arduino-pico/assets/23557839/2532c51b-9ccd-4d61-a273-6990073adc9c)
(screenshot because it's just as hard to read it as text).
So PATH environment variable is:
```
C:\\Users\\PsychoX\\.platformio\\packages\\toolchain-rp2040-earlephilhower\\bin;
C:\\Users\\PsychoX\\.platformio\\packages\\tool-rp2040tools;
C:\\Users\\PsychoX\\.platformio\\packages\\tool-openocd-rp2040-earlephilhower\\bin;
C:\\Users\\PsychoX\\.platformio\\packages\\tool-mklittlefs-rp2040-earlephilhower;
C:\\Users\\PsychoX\\.platformio\\penv\\Scripts;
C:\\Users\\PsychoX\\.platformio\\penv;
C:\\Users\\PsychoX\\.platformio\\python3;
C:\\Windows;
C:\\Windows\\System32;
// and other my system specific things
```
It seems to be GCC 12.3.0 in question
![image](https://github.com/earlephilhower/arduino-pico/assets/23557839/f6934060-8582-4f90-8b6d-9d345ae3693b)

---

But okay: I start fresh now, used PIO > Full Clean thing, also removed the downloaded package. 

My `platformio.ini`:
```
[env:pico]
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = pico
framework = arduino
board_build.core = earlephilhower
build_unflags = 
	-Wall -Wextra -Wpedantic -Wvla ; try hide warnings from framework & libraries
build_src_flags = 
	-Wall -Wextra -Wpedantic
```

My `main.cpp`:
```cpp
#include <Arduino.h>
#include <hardware/adc.h>

void setup() {  }
void loop() {  }
```

While saving the the file, it took few minutes for it all to reinstall packages. After a while I run `pio run --verbose 2>&1 > verbose.log` again, and again I see the same warnings, full log: https://gist.github.com/AgainPsychoX/914b673a1ef4828ef240c6166effa72b.

--- 

The key in the issue is that adding `-Wall -Wextra -Wpedantic` to even only my source specific build flags (via PIO `build_src_flags`) results in warnings in some headers included from my code, because (I think) those headers are included as normal project headers (`-I`/`-iwithprefixbefore`) instead as system/vendor includes (`-isystem`/`-iwithprefixs`).

I am sorry if it's frustrating, or if I am missing something. I really appreciate the time you are taking with dealing with this and many other issues.

**Reply by earlephilhower** [reply]
`-withprefixbefore` and the `-I` just set paths, they don't actually include any files unless your compilation unit includes them.

So, like I noted in the [previous comment](https://github.com/earlephilhower/arduino-pico/discussions/2104#discussioncomment-9122743), if any compilation unit includes `Arduino.h` then you're going to get those warnings with `-Wpedantic`.  You can either try and patch the SDK and the Arduino.CC upstream headers with the `#pragma`s to disable `pedantic` or not use that flag at all if you want to have clean output.

**Reply by AgainPsychoX** [reply]
> `-withprefixbefore` and the `-I` just set paths, they don't actually include any files unless your compilation unit includes them.

Yes, but they set paths to includes marked as user project files instead vendor/library files, so the warnings are propagated to them if I include them in my compilation units.

So why can't we use `-isystem`/`-iwithprefixs` instead `-I`/`-iwithprefixbefore` in `platformio-build.py` and `platform_inc.txt` of this project? Or why is it the other way currently? Maybe there is reason I am missing?

**Reply by earlephilhower** [reply]
They were chosen because they work, for ~2yrs now without incident.  Nothing really involved, they just have worked fine for everyone except this one case you're looking at AFAIK.

If you want to try changing them and make a PR, feel free, but I'm still not seeing how you're going to get any different compiler pedantic warning output.  If your own file includes any SDK header or `Arduino.h`, GCC will apply the `-Wpedantic` which is going to give you those complaints because AIUI, I don't think changing the `-iwithprefixbefore` to `-iwithprefix` would magically turn off that flag, only for an included file...

**Comment by AgainPsychoX** [comment]
I commented out any custom C++ level flag from my PlatformIO (both `-std=gnu++14 -std=gnu++17` unflags and `-std=gnu++20` added flags), I still get the similar warnings while compiling the code. 

Example part of the log:
```
[...]
Compiling .pio\build\pico\FrameworkArduino\wiring_analog.cpp.o
Compiling .pio\build\pico\FrameworkArduino\wiring_digital.cpp.o
In file included from C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/rp2_common/hardware_adc/include/hardware/adc.h:12,
                 from src\sampling.cpp:3:
C:\Users\PsychoX\.platformio\packages\framework-arduinopico/pico-sdk/src/rp2_common/hardware_gpio/include/hardware/gpio.h:881:33: warning: ISO C++ does not permit named variadic macros [-Wvariadic-macros]
  881 | #define CU_REGISTER_DEBUG_PINS(p...) \
      |                                 ^~~
Compiling .pio\build\pico\FrameworkArduino\wiring_private.cpp.o
Compiling .pio\build\pico\FrameworkArduino\wiring_pulse.cpp.o
[...]
```
